### ISSUE TITLE
Feature: do not display the actual PIN unless debug level is greater than 3

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/OpenSC#877]
**Problem:** any debugging data (`debug = 3` or higher) causes the actual value of the PIN to be logged. 

**Goal:** to have a debug setting where most everything *but the PIN* gets logged, so the log file does not need to be sanitized every time. For cases when it is uncertain whether the correct PIN is being passed on, a higher debug value (say, 8 or 9) could log the PIN too. But I'd really appreciate if there was a way to collect more data than is available at `debug = 2`, and yet not have to worry about the PIN being among that collected data.

In case it matters, that's how commercial tokends and libraries behave - they don't log the PIN itself unless specifically asked, while they allow to collect a ton of useful debugging info.

**Comment by frankmorgner** [comment]
Indeed. Could you make a PR?

**Comment by ErikHendriks** [comment]
I got the same issue, by running `pkcs15-tool -vvv --unblock-pin` it showed the puk and new pin. However it wasn't possible on all the cards.

**Comment by martinpaljak** [comment]
This issue has been on the table before, even a solution. There used to be parameter of "sensitive" logging, but that got removed, and I think for a good reason.

1) If you are paranoid, you use a pinpad reader (with a firewall!)
2) If you have a problem, it might even be in the PIN command
3) We can't control everything, a conscious effort to not log the PIN value might be meaningful, but it could "leak" in many different places, depending on drivers etc
4) Changing your PIN before sending any logs to the internet should be a sensible suggestion
5) Debug logging should normally never be turned on
6) ...

Anyway, for an APDU trace, I'd expect the FULL APDU trace, not a sanitized one. Making an effort to NOT log the plaintext PIN in full at some other stage might be meaningful though.

**Comment by frankmorgner** [comment]
OK, understood. Unless @mouse07410 or anyone else makes a PR, we're leaving this as "won't fix". (Please note that no PIN is logged when logging is disabled.)

**Comment by ErikHendriks** [comment]
Hey Martin,
The thing is I got this Belgium eID which does not reveal the pin or puk in the debug info.
Erik

-----Original Message-----

**Comment by mouse07410** [comment]
> This issue has been on the table before, even a solution. There used to be parameter of "sensitive" logging, but that got removed, and I think for a good reason.

And I respectfully disagree with the reasoning. Especially I cannot call it "good":

> 1) If you are paranoid, you use a pinpad reader (with a firewall!)

In many (if not most) cases the users do not choose a reader. And, following your psychiatric line of analogies - in my book it is clinical to demand that a user who does not want his PIN to appear in a routine debugging log must use a pinpad reader.

> 2) If you have a problem, it might even be in the PIN command

Of course, which is why the proposal was to add more gradation to the logging wrt. the PIN. Right now there are two: levels 0 through 2 log mostly nothing and omit the PIN, levels 3 through 9 log increasingly more and every one of them includes the PIN. In a sane world, probably a half of those levels (say, from 3 through 5) could do with masking the PIN - and when the efforts require PIN, the higher levels (say, from 6 through 9) would log the PIN as is, for specific debugging.

> 3) We can't control everything, a conscious effort to not log the PIN value might be meaningful, but it could "leak" in many different places, depending on drivers etc

You control your software, and should make it right. Other "many different places" is not your concern, and in my experience on Mac and Linux no "other" place popped up to spoil this picture.

> 4) Changing your PIN before sending any logs to the internet should be a sensible suggestion

I consider it not a "sensible suggestion" but an ugly workaround for what a sensible software should do, and what commercial software already does.

> 5) Debug logging should normally never be turned on

And it isn't. But sometimes one needs debugging output, which is what this issue is about.

> 6) ...

I wish you were as prompt to respond in Tokend issues, where your help could've been invaluable.

> Anyway, for an APDU trace, I'd expect the FULL APDU trace, not a sanitized one.

In most cases even in the full APDU trace you do not need to know what the exact value of the PIN was, especially if it evoked "90 00" response.

> Making an effort to NOT log the plaintext PIN in full at some other stage might be meaningful though.

I'm probably misunderstanding you. This whole issue/request was about not logging the actual value of the plaintext PIN but masking it. What "other stage" do you have in mind? Manual or scripted edit of the log file, replacing the PIN value there before submitting that log? Instead of the log-collecting software doing it right in the first place?

> Unless @mouse07410 or anyone else makes a PR, we're leaving this as "won't fix"

Understood. Disappointed. Will see what I can do.

@frankmorgner before I do any coding - would a PR that masks the actual PIN value for debug levels 3-5 be accepted? I don't want to waste efforts to just have the Martin's arguments repeated in my face.

> (Please note that no PIN is logged when logging is disabled.)

Understood. This issue was explicitly about what should be included in the **log output**, i.e., when the logging is **enabled** because the user has a need/reason to log the processing.

**Comment by martinpaljak** [comment]
@mouse07410 I understand you perfectly, and as I don't do that much practical coding on OpenSC these days, feel free to pay no attention to my comments. I think we have a different background and context: as long as someone is using a PC (which IMHO is the main target for OpenSC) in my region it pretty much means that the user is in control of both hardware as well as software (yes, you have builtin laptop readers, but you still have a USB port). You can get a firewalling pinpad reader for 20€ from almost any IT shop in Estonia.

My logic goes via a different route. Once you're able to change the intentional debugging flag, you're probably able do other things on the system as well. Like run malware. Or sniff other applications than OpenSC. If you don't have a realistic threat, and realistic countermeasures, you are dealing with snakeoil.

I don't mean that we should not make efforts to make things harder for bad guys, but at some point there's a balance somewhere between effort and payoff. I think the balance of "not having debugging turned on" is a justified one, compared to a lot of effort for trying to mask potential leaks without a guaranteed outcome. KISS.

The whole debuggin layer of OpenSC is pretty arbitrary, the levels have (AFAIR) never been even semi-formally specified, so whatever adjustments you make, there will be some work to make sure that none of the codebase violates the new rules. 

The other thing is having total understanding of the codebase, to be *sure* that you have covered all plaintext (eg code paths that precede APDU logging) logging cases. And maintain that with future code reviews. And then filter APDU-s. And then fail when some weird (*cough-cough, no names given*) card decides to send PIN codes via some custom APDU path. True, a firewalling pinpad reader would also fail in the case when non-standard APDU-s would be used.

The whole point: feel free to make a PR. But I doubt that it would cover all realistic cases for possible PIN leakage, without adding logical and structural complexity that would better be avoided by just declaring debugging a sensitive operation.

**Comment by mouse07410** [comment]
> as long as someone is using a PC (which IMHO is the main target for OpenSC) in my region it pretty much means that the user is in control of both hardware as well as software

First, this is incorrect regarding employees - they may or may not be able to choose either hardware or software. In some organizations only certain readers are allowed. Some may
have lists of permitted software.

Second - even if a user is in control, there may be reasons precluding him from choosing the kind of reader that you prefer. For example, I use Yubikeys - and they can't be used with a pinpad reader. Others use smartcard-hsm that can't be used with a pinpad reader either.

> You can get a firewalling pinpad reader for 20€ from almost any IT shop in Estonia

First, I may not be able/allowed to buy hardware from Estonia. Second, some of my tokens (like the two above) are USB-direct, and can't work with that reader. Third, I personally don't want to waste 20€ on a feature that does not address my needs (including portability).

> Once you're able to change the intentional debugging flag, you're probably able do other things on the system as well. Like run malware. Or sniff other applications than OpenSC. 

Of course.

> If you don't have a realistic threat, and realistic countermeasures, you are dealing with snakeoil

I'm sorry for not explaining my purpose here. The goal of PIN hiding is not some great Data Leak Prevention, or such. It's merely to make it easier on a developer or user, so he does not have to waste time sanitizing the log before sending it out.

>  don't mean that we should not make efforts to make things harder for bad guys, but at some point there's a balance somewhere between effort and payoff

It's not about "making things harder for bad guys", it's about making things a little easier for good guys, taking a bit of work off their timetable.

> The whole debuggin layer of OpenSC is pretty arbitrary, the levels have (AFAIR) never been even semi-formally specified, so whatever adjustments you make, there will be some work to make sure that none of the codebase violates the new rules

OK, understood.

> The other thing is having total understanding of the codebase, to be sure that you have covered all plaintext (eg code paths that precede APDU logging) logging cases. And maintain that with future code reviews. And then filter APDU-s.

Understood.

> And then fail when some weird (`cough-cough, no names given`) card decides to send PIN codes via some custom APDU path.

I'm rather pragmatic here. As long as *my* cards (CAC, PIV, Yubikey, a couple of others) are OK, it's fine with me if some weird card does leak its PIN somehow. I let it be somebody else's problem.

> The whole point: feel free to make a PR. But I doubt that it would cover all realistic cases for possible PIN leakage, without adding logical and structural complexity that would better be avoided by just declaring debugging a sensitive operation.

Thanks. I'll see what I can do PR-wise. As for the goal - it's far more modest: not to cover all realistic cases for possible PIN leakage, but making sure the "normal" currently-exhibited PIN leakage paths are debug level-controlled, mainly for the comfort of those who need to produce those logs and turn them over to somebody else. That's all.
